src/chainbin/nnet3-chain-combine.cc

```diff
@@ -102,29 +168,47 @@ int main(int argc, char *argv[]) {
       KALDI_ASSERT(!egs.empty());
     }
 
+    // first evaluates the objective using the last model.
+    int32 best_num_to_combine = 1;
+    double best_objf = ComputeObjf(batchnorm_test_mode, dropout_test_mode,
+        egs, moving_average_nnet, chain_config, den_fst, &prob_computer);
+    KALDI_LOG << "objective function using the last model is " << best_objf;
 
     int32 num_nnets = po.NumArgs() - 3;
-    NnetChainCombiner combiner(combine_config, chain_config,
-                               num_nnets, egs, den_fst, nnet);
-
+    // then each time before we re-evaluate the objective function, we will add
+    // num_to_add models to the moving average.
+    int32 num_to_add = (num_nnets + max_objective_evaluations - 1) /
+                       max_objective_evaluations;
     for (int32 n = 1; n < num_nnets; n++) {
       std::string this_nnet_rxfilename = po.GetArg(n + 2);
       ReadKaldiObject(this_nnet_rxfilename, &nnet);
-      combiner.AcceptNnet(nnet);
+      // updates the moving average
+      UpdateNnetMovingAverage(n + 1, nnet, &moving_average_nnet);
+      // evaluates the objective everytime after adding num_to_add model or
+      // all the models to the moving average.
+      if ((n - 1) % num_to_add == num_to_add - 1 || n == num_nnets - 1) {
+        double objf = ComputeObjf(batchnorm_test_mode, dropout_test_mode,
+            egs, moving_average_nnet, chain_config, den_fst, &prob_computer);
+        KALDI_LOG << "Combining last " << n + 1
+                  << " models, objective function is " << objf;
+        if (objf > best_objf) {
+          best_objf = objf;
+          best_nnet = moving_average_nnet;
+          best_num_to_combine = n + 1;
+        }
+      }
     }
+    KALDI_LOG << "Using the model averaged over last " << best_num_to_combine
```

Comment:
Please make sure that steps/info/chain_dir_info.pl is modified so that it can read both the output of this binary, and the old one.  The same goes for nnet3_dir_info.pl.   They print something like "combine: xxx->yyy".  And for the new one, you can make it print "combine: xxx->yyy (over n)".